### PR TITLE
#495 Add x/act integration test for adding/removing owners

### PR DESCRIPTION
**Pull request by backsapc**
1. Fixed import of spaces from Genesis.
1. Generated a new snapshot with many users.
1. Implemented an integration test.

<s>Notes: I found out that the rule must be in the format `any(2, warden.space.owners)` (including the `warden.` prefix). That wasn't obvious, even after reading the documentation.</s>

https://github.com/warden-protocol/wardenprotocol/issues/495

## Summary by CodeRabbit

- **New Features**
	- Introduced new target `snapshot-many-users` for generating snapshots in blockchain applications with multiple users.
	
- **Enhancements**
	- Improved the import process in the `Keeper` struct with updated state handling.
	- Added new methods for retrieving and caching addresses in the `Wardend` struct.

- **Testing**
	- Added comprehensive test cases for owner approval actions, covering user interactions and validations.

### CODE REVIEW

**Comment by mn13** [comment]
> after reading the documentation

Does format mentioned somewhere in docs? If it is so, I think it better to edit it then

File: tests/cases/owner_approve_action.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+	resApproveBob := alice.Tx(t, "act approve-action --action-id 1")
+	checks.SuccessTx(t, resApproveBob)
+	client.EnsureSpaceAmount(t, ctx, users.Bob.Key, 1)
+
+	resNewRule := alice.Tx(t, "act new-rule --name approve_requires_two --definition \"any(2, warden.space.owners)\"")
+	checks.SuccessTx(t, resNewRule)
+
+	resAssigneRule := alice.Tx(t, "warden new-action update-space --space-id 1 --admin-rule-id 1")
+	checks.SuccessTx(t, resAssigneRule)
```
Comment (mn13) [inline tests/cases/owner_approve_action.go]: Suggest to add check.FailTx on new-action from non-owner

**Review by mn13 — CHANGES_REQUESTED** [review]

**Comment by backsapc** [comment]
> > after reading the documentation
> 
> Does format mentioned somewhere in docs? If it is so, I think it better to edit it then

image

It looks like my bad actually..

File: tests/cases/owner_approve_action.go (hunk 1 above)
Comment (backsapc) [reply tests/cases/owner_approve_action.go]: ✅

**Review by mn13 — DISMISSED** [review]

File: warden/x/warden/keeper/genesis.go [hunk 2]
```diff
@@ -22,6 +22,13 @@ func (k *Keeper) ImportState(ctx sdk.Context, genState types.GenesisState) error
 		return fmt.Errorf("failed to import spaces: %w", err)
 	}
 
+	for _, key := range genState.Spaces {
+		err := k.SpacesKeeper.updateSpaceOwners(ctx, key, nil)
+		if err != nil {
+			return fmt.Errorf("failed to import space owners: %w", err)
+		}
+	}
```
Comment (pitasi) [inline warden/x/warden/keeper/genesis.go]: maybe one thing we could do is to put this inside a new method `k.SpacesKeeper.Import(...)` that inside calls keeps the index updates.

so we don't leak the spaceOwners index here

**Review by pitasi — CHANGES_REQUESTED** [review]
One problem I see here are the hardcoded addresses (in the code they're called "key" but they're actually the address). They would change every time we regenerate the snapshots. To avoid this problem I suggest we do this: we add a new method to `warden_cli.go`:

```go
type Wardend struct {
	BinPath     string
	Name        string
	QueryAppend string
	TxAppend    string
	Node        *WardenNode

	address string
}

// fix NewWardend to init the new fields: Name and Node

func (cli *Wardend) Address(t *testing.T) string {
	if cli.address != "" {
		return cli.address
	}

	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()

	cmd := &Exec{
		Bin:    "sh",
		Args:   []string{"-c", fmt.Sprintf("%s keys show %s -a --keyring-backend test --keyring-dir %s", cli.BinPath, cli.Name, cli.Node.Home)},
		Stdout: &iowriter.IOWriter{},
		Stderr: &iowriter.IOWriter{},
	}
	err := cmd.Run(ctx)
	require.NoError(t, err)

	output := cmd.Stdout.String()
	cli.address = strings.TrimSpace(output)

	return cli.address
}
```

so the `TestUsers` and `User` structs become unnecessary and can be deleted.

We can use `alice := exec.NewWardend(c.w, "alice")` directly, and then we can `alice.Address(t)` to get the address for alice, when needed.

**Comment by pitasi** [comment]
> It looks like my bad actually..

i agree that it's not particularly clear from the docs how to use shield in general, and what variables are available, we'll need to improve it

File: warden/x/warden/keeper/genesis.go (hunk 2 above)
Comment (backsapc) [reply warden/x/warden/keeper/genesis.go]: Thank you for your comments, changed as requested

**Review by pitasi — APPROVED** [review]
thanks, looks good to me!

please merge it using "squash and merge", or alternatively you can manually edit your commits (I think the minimum number is two: one for fixing genesis import, one for adding test+snapshot) and then you can "rebase and merge" 🙌

**Comment by backsapc** [comment]
> thanks, looks good to me!
> 
> please merge it using "squash and merge", or alternatively you can manually edit your commits (I think the minimum number is two: one for fixing genesis import, one for adding test+snapshot) and then you can "rebase and merge" 🙌

It looks like squash isn't enabled here
image

**Pull request merged**